backend/CircleForms/Services/Database/RedisCacheRepository.cs

```diff
@@ -0,0 +1,225 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using CircleForms.Models;
+using CircleForms.Models.Posts;
+using CircleForms.Services.Database.Interfaces;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using StackExchange.Redis;
+
+namespace CircleForms.Services.Database;
+
+public class RedisCacheRepository : ICacheRepository
+{
+    private const string _postsSet = "posts";
+    private const string _activeSet = "posts:active";
+    private const string _inactiveSet = "posts:inactive";
+    private const string _pinnedSet = "posts:pinned";
+    private const string _userIds = "user_ids";
+
+    private readonly ILogger<RedisCacheRepository> _logger;
+    private readonly IMapper _mapper;
+    private readonly string[] _postOccupation = {_postsSet, _activeSet, _pinnedSet, _inactiveSet};
+    private readonly IDatabase _redis;
+
+    public RedisCacheRepository(IConnectionMultiplexer redis, IMapper mapper, ILogger<RedisCacheRepository> logger)
+    {
+        _mapper = mapper;
+        _logger = logger;
+        _redis = redis.GetDatabase();
+    }
+
+    public async Task IncrementAnswers(string id)
+    {
+        await _redis.StringIncrementAsync(ToPostAnswersCount(id));
+    }
+
+    public async Task<bool> PinPost(string id)
+    {
+        var key = ToPostId(id);
+
+        if (!_redis.KeyExists(key))
+        {
+            return false;
+        }
+
+        await _redis.SetAddAsync(_pinnedSet, key);
+
+        return true;
+    }
+
+    public async Task<bool> UserExists(string id)
+    {
+        return await _redis.SetContainsAsync(_userIds, id);
+    }
+
+    public async Task AddUserToUserIds(string id)
+    {
+        await _redis.SetAddAsync(_userIds, id);
+    }
+
+    public async Task Unpublish(string id)
+    {
+        await PurgePost(id);
+    }
+
+    public async Task<PostRedis> Publish(Post post)
+    {
+        await PurgePost(post.ID); //Purge just for fun
+
+        var postId = ToPostId(post.ID);
+        var postRedis = _mapper.Map<PostRedis>(post);
+        var postJson = JsonConvert.SerializeObject(postRedis);
+
+        _logger.LogInformation("Adding {Post} to the cache", postId);
+        if (!await _redis.StringSetAsync(postId, postJson))
+        {
+            _logger.LogError("Could not post {@Post} to the redis cache", postRedis);
+
+            return null;
+        }
+
+        var publishUnixTime = post.PublishTime.ToUnixTimestamp();
+        _logger.LogInformation("Adding {PostId} to the cached posts set", postId);
+        if (!await _redis.SortedSetAddAsync("posts", postId, publishUnixTime))
+        {
+            _logger.LogError("Could not post {PostId} with {UnixTime} to the redis posts", postId, publishUnixTime);
+        }
+
+        await SetActivity(post);
```

Comment:
This should only be done on `post.Accessibility == Accessibility.Public`, otherwise it's possible to see hidden posts when filtering

Reply:
it's already checks it on the `PostsService` layer